README.md
![](docs/logo.png)

Gluon
=============
We can call gluon a project boss. Simply manages routes, decides which route should go first, also manages models and database connection. With gluon, project development became so much easier.. 

installation
-----------------

add dependency `gluon` to  dependencies or type console

```bash
npm install gluon --save
```


basic usage
----------------

```javascript
const gluon = require('gluon');
const app = gluon();

app.listen(80);
```

use with external app
```javascript
const gluon = require('gluon');
const express = require('express');

const app = express();
gluon({ app });

app.listen(80);
```

using cwd + sub location. (Note if you make your project into src directory, you need this)
```javascript
const gluon = require('gluon');
const app = gluon({dir: 'src/'}); // app.js located as cwd/src/app.js

app.listen(80);
```
routing mechanism
--------------------

create a directory that called by `routes`

routes/home.js
```javascript
const gluon = require('gluon');
const router = gluon.router();

router.get('/', (req, res) => {
  res.send("ok");
});

module.exports = router;
```

app.js
```javascript
const gluon = require('gluon');
const app = gluon({dir: 'src/'}); // app.js located as cwd/src/app.js

// nothing required just works
app.listen(80);
```

try to connect `http://localhost:80/home`

sub-routing and index
------------------------


create a directory that called by `routes`

routes/home.js
```javascript
const gluon = require('gluon');
const router = gluon.router();

router.get('/', (req, res) => {
  res.send("ok");
});

module.exports = router;
```

routes/sub/index.js
```javascript
const gluon = require('gluon');
const router = gluon.router();

router.get('/', (req, res) => {
  res.send("sub");
});

module.exports = router;
```


routes/sub/home.js
```javascript
const gluon = require('gluon');
const router = gluon.router();

router.get('/', (req, res) => {
  res.send("sub-> home");
});

module.exports = router;
```

app.js
```javascript
const gluon = require('gluon');
const app = gluon({dir: 'src/'}); // app.js located as cwd/src/app.js

// nothing required just works
app.listen(80);
```

try to connect these

* `http://localhost:80/home` -> "ok"
* `http://localhost:80/sub` -> "sub"
* `http://localhost:80/sub/home` -> "sub-> home"

> **Note:** Gluon has order mechanism that always index is last in own group. 
> **Example:** 
> * Next most powered group ordered by alphabet `A/B/C` `A/B/C.js`
> * Next most powered group and index `A/B` `A/B/index.js`
> * index  `/` `index.js`

database and models
--------------------
Gluon uses sequelize to communicate with database. But first you must create `config` folder in your cwd. For example your cwd is `/root/myproject/`, and your project files in `/root/myproject/src` create config in `/root/myproject/config` and don't forget to add dir option to gluon's first parameter.

> **Note:** These modules requires `sequelize` and `config` please install before using it

you can create a json file that called by `default.json`

default.json
```json
{
  "database": {
    "dialect": "mysql",
    "host": "your host address",
    "user": "database user name",
    "database": "database name",
    "password": "database password"
  }
}
```

now create a folder in src folder that called by `models`

models/user.js
```javascript
const Sequelize = require('sequelize');
const db = require('gluon/db');

module.exports = db.define('User', {
  account: {
    type: Sequelize.STRING(32),
    allowNull: false,
    unique: true,
    validate: {
      isAlpha: true
    }
  },

  password: {
    type: Sequelize.STRING(32),
    allowNull: false,
    validate: {
      is: /^[a-f0-9]{32}$/
    }
  },

}, {
  freezeTableName: true,
  paranoid: true
});
```

everything is ready, start the project. gluon will connect db automaticly and load your models immediately. use your models in router

routes/user.js
```javascript
const gluon = require('gluon');
const router = gluon.router();
const user = require('../models/user');

router.get('/', (req, res) => {
    user.all().then((data) => res.json(data));
});

module.exports = router;
```


Other awesome features
----------------------

### Generic option (default: true)

if you set generic option, you get plenty of new response types.
```javascript
app.use((req, res) => {
    res.ok("ok :)");
});
```

* **ok** _Everything is OK_ 200 (text or json) 
* **notFound** _Can't find user or something_ 404 (info) `{error: true, info}`
* **badRequest** _You made wrong request_ 400 (requiredFields) `{error: true, requiredFields}`
* **validation** _Your request didn't passed the validation_ 400 (type, fields) `{error: true, type, fields}`
* **unauthorized** _You are trying to use service without authorization_ 401 (info) `{error: true, info}`
* **redirectRequest** _Redirects a request to another one, like use POST please_ (info) `{error: true, info}`
* **unknown** _Bad Error_ 500 (info) `{error: true, info}`
* **database** _Database error (if caused by validation then it automaticly calls)_ 500 (err) `{error: true, info: 'Database triggered an error. Please check your request. If there is no problem then contact with service provider.'}`
* **expiredToken** _Use when a token expires_ 408 (info) `{error: true, info}`

### Ready option and Before option
use when you need to wait all gluon initial job. Its very handy option

```javascript
const gluon = require('gluon');
const app = gluon({
    dir: 'src/',
    before: (app) => {
        app.use(require('compression')());
    },
    ready: (app) => {
        app.use((err, req, res, next) => {
            res.send("something wrong");
        });
    }
});

app.listen(80);
```

### Gluon request body controller

It helps you when you need control body. Lets check examples

> **Note:** Gluon request body controller requires **generic=true**

routes/user.js
```javascript
const gluon = require('gluon');
const control = require('gluon/control');
const router = gluon.router();
const user = require('../models/user');

// just give model
router.get('/', control(user), (req, res) => {
    user.all().then((data) => res.json(data));
});


// or give an array
router.get('/other', control(['account', 'password']), (req, res) => {
    
});

// if you don't want to control id
router.get('/', control(user, false), (req, res) => {
    user.all().then((data) => res.json(data));
});


module.exports = router;
```

If request cannot pass body controller, it send badRequest.

### Gluon generator

If you bored to create CRUD you can use gluon generator.

> **Note:** This generator requires `js-md5` please install before using it. With this you don't need to serialize `password` field. it does automaticly.
>
> **Warning:** Before generator you must definitely do authentication. 

supported routes
```
GET /
GET /all
GET /count
POST /all (filter)
POST /count (filter)
GET /:id
DELETE /:id
POST / (create new model)
PATCH /:id (update exiting model)
```

#### Example 

routes/user.js
```javascript
const gluon = require('gluon');
const generator = require('gluon/generator');
const router = gluon.router();
const user = require('../models/user');

generator(router, user);

module.exports = router;
```

### Gluon logger

Starting from 1.1.0 gluon no longer uses `ezlogger`, we suggest to use `gluon/logger`.

#### Options

These are default values of options, you can change them using by `logger.set( )` or you can install `config` module and create `logger` object (example 2)

```javascript
{
  dir: './logs',
  level: 'DEBUG',
  fileFormat: 'log.{date}.txt',
  dateFormat: '{year}/{month p 2}/{day p 2}',
  fileDateFormat: '{year}.{month p 2}.{day p 2}',
  timeFormat: '{hour p 2}:{minute p 2}:{second p 2}',
  type: 'full',
  full: '{date} {time} {type} {file}:{line} {message}',
  exceptDate: '{time} {type} {file}:{line} {message}',
  simple: '{type} {message}',
  withFile: '{type} {file}:{info} {message}',
  request: '[{req.method}] {coloredStatus} {time} {req.headers.host}{req.url} from {req.ip}'
}
```

#### Create new logging type

```javascript
const logger = require('gluon/logger');
logger.set('myOwnLoggerType', '{time} {file}:{line} {type} {message}');
logger.set('type', 'myOwnLoggerType');
```

#### Example

app.js
```javascript
const gluon = require('gluon');
const logger = require('gluon/logger')
logger.set('type', 'full'); // logs everything.. default value: full
logger.set('dir', './logs/'); // cwd/logs

const app = gluon({log: true}); // log route actions too
app.listen(80);
logger.log('server started to listen at 80');
```

routes/home.js
```javascript
const gluon = require('gluon');
const logger = require('gluon/logger');
const router = gluon.router();


module.exports = router;
```

#### Example 2: with config

> **Note:** You must install config first. `npm install config --save`

app.js
```javascript
const gluon = require('gluon');
const logger = require('gluon/logger');

const app = gluon();
app.listen(80);
logger.log('server started to listen at 80');
```

routes/home.js
```javascript
const gluon = require('gluon');
const logger = require('gluon/logger');
const router = gluon.router();


module.exports = router;
```

config/default.json
```json
{
    "logger": {
        "level": "LOG",
        "type": "myOwn",
        "myOwn": "{type} {file}:{line} {message}"
    }
}
```


### Gluon defaults by config

config/default.json
```js
{
    "gluon": {
        "generic": "false", // don't use generic system
        "log": "false", // don't log requests
        "dir": "./src" // where is my app.js from cwd
    }
}
```

### Direct listen from code or config

> **Note:** if PORT is given by environment then this port is not will be evaluated

config/default.json
```javascript
{
    "gluon": {
        "listen": 80 
        // or
        "listen": {
            "ip": "1.2.3.4",
            "port": 80
        }
    }
}
```

or

```javascript
const gluon = require('gluon');
const app = gluon({
    listen: {
        ip: "127.0.0.1",
        port: 80
    }
});
```

### Create public folders

> **Note:** default public folder is `./public` (searching as `cwd/project dir/public`)

config/default.json
```javascript
{
    "gluon": {
        "publicSource": "./static"
        // or
        "publicSource": [
            "./public", // public has more priority than static one
            "./static"
        ]
    }
}
```

or

```javascript
const gluon = require('gluon');
const app = gluon({
    publicSource: './public',
});
```


### Ignore a route

You might want to ignore a route that shouldn't routed by gluon itself. Maybe you want to use this for other purposes.

routes/authentication.js
```javascript
const gluon = require('gluon');
const router = gluon.router(null, true);

router.use((req, res, next) => {
    if(authentication check..) return res.unauthorized('you must log in'); 
    next();
});

module.exports = router;
```

routes/logout.js
```javascript
const gluon = require('gluon');
const router = gluon.router();

router.use(require('./authentication'));
router.all('/', (req, res) => {
    logout..
});

module.exports = router;
```


### Authentication and role management
With gluon 1.3.0 token based authentication generator released. Generator uses sequelize to communicate database. Creates 2 tables, "Role" and "Token".
Before doing anything, make database connection correctly. Lets look properties

config/default.json
```javascript
{
    "gluon": {
        "auth": {
            "base": "token"
        }
    }
}
```

gluon/auth only works if you set base = "token"

config/default.json
```javascript
{
    "gluon": {
        "auth": {
            "base": "token",
            "model": "user", // which model is owner of token. Make sure your model is exist in your models folder.
            "disable": false, // default= false, if you just disable authentication for some purpose, set this to true
            "expire": 43200, // default=43200, how much seconds will be need for token to expire?
            "allow": [
                "/authentication/does/not/exist/in/this/path",
                "/login", // default you can't block /login
                "/register", // same here these are default.
                "/dashboard",
                "/blabla/:id", // you can do this too. but only format is ":xxxx" nothing more, simply makes "[^\/]*" regex
                "regexp:^/damn/you/know+/regexp" // you can use regexp with this "regexp:" prefix
            ],
            "routes": {
                "/admin": "admin", // simply /admin for only "admin" role owners
                "/see": "canSee",
                "/see/important": "important", // this route will require both "canSee" and "important" roles
                "/edit/:id": "canEdit"
            },
            "token": "./token", // If you don't know what are you doing, Don't change this. This setting simply changes the token model
            "role": "./role" // Same as here if you don't know bla bla.. Don't change it..
        }
    }
}
```

Thats enough for configurations. Simply routes make static blocks. If you want to dynamic blocks you must use `req.auth`

* `req.auth.login(model)`: creates token for model
* `req.auth.logout()`: removes token
* `req.auth.hasRole(role)`: checks for role
* `req.auth.addRole(role)`: adds role
* `req.auth.removeRole(role)`: removes role
* `req.auth.hasRoles(roles)`: checks for roles
* `req.auth.addRoles(roles)`: adds roles
* `req.auth.removeRoles(roles)`: removes roles

routes/login.js
```javascript
const gluon = require('gluon');
const router = gluon.router();
const User = require('../models/user');

router.get('/', (req, res) => {
    // controls etc..
    // fetch model that defined in "gluon.auth.model"
    User.find({
        where: {}
    }).then((user) => {
        req.auth.login(user).then((token) => {
            res.ok(token.code);
        });
    });
});

module.exports = router;
```

routes/someServiceThatHasAuthentication.js
```javascript
// note this route isn't allowed directly. Directly allowing means closing authentication for that route.
const gluon = require('gluon');
const router = gluon.router();

router.get('/:id', (req, res) => {
    // lets say i need to check some dynamic role "canEditDoor5"
    req.auth.hasRole("canEditDoor" + req.params.id).then((has) => {
        // if user has canEditDoor5 then has will be true
    });

    // you can add these roles by using some database tools or direct `req.auth.addRole`
    req.auth.addRole("canEditDoor" + req.params.id).then((role) => {

    });

    // if you want to change some user's role then
    const Role = require('gluon/role'); // gather role model
    Role.addRole(id of user, "canEditDoor5").then((role) => {

    });
});

module.exports = router;
```

By time tokens will stay there. You must use `node-schedule` or similar cron job tools. You can use this.

app.js
```javascript
const gluon = require('gluon');
const schedule = require('node-schedule');
const app = gluon({
  ready: () => {
    const Token = require('gluon/token');
    schedule.scheduleJob('*/10 * * * *', () => { // remove expired tokens for every 10 minutes
      Token.destroy({
        where: {
          expire: {
            $lt: new Date
          }
        }
      });
    });
  }
});
```


### Parameter routes

If you want to use parameter but you don't want to create single file. You can use `@name` syntax for folders. for ex:

```javascript
/routes/index.js -> /
/routes/@id/index.js -> /:id
/routes/dashboard/@dashboard_id/index.js -> /dasboard/:dashboard_id
/routes/fridge/@id/index.js -> /fridge/:id
/routes/fridge/@id/dashboard.js -> /fridge/:id/dashboard
/routes/fridge/@id/temperature/@temperature_id.js -> /fridge/:id/temperature/:temperature_id
```

### Logger special variables

app.js

```javascript
var logger = require('gluon/logger');
var cluster = require('cluster');
logger.set('worker_id', cluster.worker ? cluster.worker.id : '-');
```

config

```javascript
{
  "logger": {
    "level": "LOG",
    "type": "clusterLogger",
    "clusterLogger": "WORKER {var.worker_id} {date} {time} {type} {file}:{line} {message}"
  }
}
```


changelog.txt
# Changelog

#### 1.2.8
* Fixed redefinition of colored prototypes.

#### 1.2.9
* Database config now has all properties. Not just login credentials. (ex: timeout)

#### 1.3.0
* Build-in Token Based Authentication
* Build-in Role System

#### 1.3.1
* Role system has very tiny bug, we fixed it.

#### 1.4.0
* Routes now support "@name". You can make parameters with it. Ex: "./routes/fridge/@id/dashboard.js" -> "/fridge/:id/dashboard"

#### 1.4.1
* Fixed route index issue.

#### 1.4.3
* Added mergeParams option to gluon.router

#### 1.4.4
* If there is no role required then, just accept request directly, no need to use database for role checking at this moment.

#### 1.5.0
* Added hasRoles, addRoles, removeRoles functions to role management.

#### 1.5.1
* Public sources should not require authentication. We allowed them now.

#### 1.5.2
* License added to folder.

#### 1.6.0
* token-redis authentication support

#### 1.7.0
* support node 4.x

#### 1.7.1
* logger now has "var" variable.

#### 1.7.2
* fixed deprecation warning error

#### 1.7.3
* fixed gluon/control's control mechanism on nullable fields.

#### 1.7.4
* fixed gluon/control's added order for /all

#### 1.7.5
* fixed gluon/generator's create, userId auto complete bug.

#### 1.8.0
* added recursive search for models in generator

#### 1.8.1
* fixed recursive search.

#### 1.8.2
* fixed recursive search.

#### 1.8.3
* fixed recursive search.

#### 1.8.4
* token and role now has camelcase names


gluon.js
var express = require('express');
var os = require('os');
var path = require('path');
var ezFormat = require('ezformat');
var bodyParser = require('body-parser');
var requireDir = require('./utils/require-dir');
var routeLoader = require('./utils/route-loader');
var logger = require('./logger');

var defaults = {
  generic: true,
  log: true,
  publicSource: './public',
  models: './models',
  routes: './routes',
  dir: '',
  auth: null
};

var authDefaults = {
  disable: false,
  expire: 43200,
  model: "user",
  token: './token',
  role: './role'
};

try {
  var config = require('config');
  var gluonDefaults = config.get('gluon');
  Object.assign(defaults, gluonDefaults);
} catch (e) {
  // if there is no config installation than ignore
  logger.debug('config package doesn\'t exists.');
}

/**
 * Gluon main function
 * @function
 * @param {Object} [options] Gluon startup options
 * @param [options.app] Express application container. if you make this empty then system create new one.
 * @param {Boolean} [options.generic=true] Should gluon initialize generic extension.
 * @param {Boolean} [options.log=true] Should gluon initialize request logging mechanism.
 * @param {Function(app|*, Logger)} [options.ready] Ready signal
 * @param {Function(app|*, Logger)} [options.before] Before anything loaded
 * @param {String} [options.dir=''] Location of project
 * @param {String} [options.models='./models'] Location of models folder
 * @param {String} [options.routes='./routes'] Location of routes folder
 * @param {String} [options.auth=null] Authentication controller
 * @param {String|Array<String>} [options.publicSource='./public'] Location of project
 * @param {{ip: String, port: Number}|Number} [options.listen] Should i listen?
 * @returns {app|*}
 */
function Gluon(options) {
  (options != undefined || (options = {}));
  Object.keys(defaults).forEach(function (key) {
    if (options[key] == undefined) options[key] = defaults[key];
  });

  var app = options.app || express();

  logger.debug('Initializing...');
  if (options.generic == true) {
    logger.debug('Request generic extension loading..');
    require('./utils/generic-response')(app);
  }

  if (options.before) options.before(app, logger);

  logger.debug('Body parser loading..');
  app.use(bodyParser.urlencoded({extended: false}));
  app.use(bodyParser.json());

  if (options.log == true) {
    logger.debug('Request logging mechanism generating..');
    app.use(function (req, res, next) {
      var now = new Date;

      if (logger.level() <= 0) {
        var keys = Object.keys(req.body);
        if (keys.length > 0) logger.debug('body information ({length}): {2.EOL}{1 j 4}', keys, req.body, os);
      }

      res.on('finish', function () {
        var status = res.statusCode + '';
        var coloredStatus = res.statusCode >= 500 ? status.red : res.statusCode >= 400 ? status.red
          : res.statusCode >= 300 ? status.yellow : res.statusCode >= 200 ? status.green : status;
        var time = new Date - now;
        time = time < 100 ? ((time >>> 0) + 'ms').green
          : time < 500 ? ((time >>> 0) + 'ms').yellow
          : time < 1000 ? ((time >>> 0) + 'ms').red
          : time < 60000 ? ('{0:fixed:2}s'.format(time / 1000)).red
          : time < 3600000 ? ('{0:fixed:2}m'.format(time / 60000)).red
          : ('{0:fixed:2}h'.format(time / 3600000)).red;

        logger.log(logger.get('request'), {req: req, coloredStatus: coloredStatus, time: time});
      });
      next();
    });
  }


  if (options.publicSource) {
    if (typeof options.publicSource == 'string') {
      var location = path.resolve(process.cwd(), options.dir, options.publicSource);
      app.use(express['static'](location));
      logger.debug('Folder {0} has been set as public', location);
    } else if (options.publicSource == null) {
      logger.debug('There is no public folder');
    } else if (options.publicSource.constructor == Array) {
      options.publicSource.forEach(function (source) {
        var location = path.resolve(process.cwd(), options.dir, source);
        app.use(express['static'](location));
        logger.debug('Folder {0} has been set as public', location);
      });
    }
  }

  if (options.auth && options.auth.base == "token") {
    options.auth = Object.assign(authDefaults, options.auth);
    if (options.auth.disable == false || options.auth.disable == undefined || options.auth.disable == "false") {
      var authLocation = path.resolve(process.cwd(), options.dir, options.models, options.auth.model);
      try {
        global._gluon_auth_model = require(authLocation);
        global._gluon_auth_expire = options.auth.expire;
      } catch (e) {
        logger.error('Auth model is not exist in {0}', authLocation);
      }

      if (global._gluon_auth_model) {
        var Token = require(options.auth.token);
        var Role = require(options.auth.role);

        //noinspection JSUnresolvedVariable
        var allow = options.auth.allow ? options.auth.allow.map(function (route) {
          return new RegExp(route.indexOf("regexp:") == 0 ? route.substring(7) : "^" + route.replace(/:[^\\/]+/g, '.*'));
        }) : [];

        var routes = Object.keys(options.auth.routes || {}).map(function (route) {
          return {
            match: new RegExp(route.indexOf("regexp:") == 0 ? route.substring(7) : "^" + route.replace(/:[^\\/]+/g, '.*')),
            role: options.auth.routes[route]
          };
        });

        app.use(function (req, res, next) {
          //noinspection JSValidateJSDoc
          /**
           * Authentication protocol for gluon
           */
          req.auth = {
            /**
             * Creates a token for model
             * @param {Model} model
             * @returns {Promise.<TResult>}
             */
            login: function (model) {
              return Token.create({
                code: Token.generateCode(),
                expire: Token.defaultExpire(),
                ownerId: model.id
              }).then(function (data) {
                return data;
              }).catch(function (err) {
                res.database(err)
              });
            },


            /**
             * Removes token from owner
             * @returns {Promise.<TResult>}
             */
            logout: function () {
              return Token.destroy({
                where: {
                  code: req.get('token')
                }
              }).then(function (data) {
                return data;
              }).catch(function (err) {
                res.database(err)
              });
            },


            /**
             * Adds a role to owner
             *
             * @param {String} role Which role
             * @returns {Promise.<Instance>}
             */
            addRole: function (role) {
              return Role.findOrCreate({
                where: {
                  code: role,
                  ownerId: req[options.auth.model].id
                },

                defaults: {
                  code: role,
                  ownerId: req[options.auth.model].id
                }
              }).spread(function (role, created) {
                return role
              }).catch(function (err) {
                res.database(err)
              });
            },


            /**
             * Adds a roles to owner
             *
             * @param {Array<String>} roles Which roles
             * @returns {Promise.<Instance>}
             */
            addRoles: function (roles) {
              roles.map(function (role) {
                return Role.findOrCreate({
                  where: {
                    code: role,
                    ownerId: req[options.auth.model].id
                  },

                  defaults: {
                    code: role,
                    ownerId: req[options.auth.model].id
                  }
                }).spread(function (role, created) {
                  return role
                }).catch(function (err) {
                  res.database(err)
                });
              });

              return Promise.all(roles);
            },


            /**
             * Removes a role from owner
             *
             * @param {String} role Which role
             * @returns {Promise.<Number>}
             */
            removeRole: function (role) {
              return Role.destroy({
                where: {
                  code: role,
                  ownerId: req[options.auth.model].id
                },
                limit: 1
              }).catch(function (err) {
                res.database(err)
              });
            },

            /**
             * Removes a role from owner
             *
             * @param {Array<String>} roles Which roles
             * @returns {Promise.<Number>}
             */
            removeRoles: function (roles) {
              return Role.destroy({
                where: {
                  code: {
                    $in: roles
                  },
                  ownerId: req[options.auth.model].id
                }
              }).catch(function (err) {
                res.database(err)
              });
            },


            /**
             * Checks for role
             *
             * @param {String} role Which role
             * @returns {Promise.<Boolean>}
             */
            hasRole: function (role) {
              return Role.count({
                where: {
                  code: role,
                  ownerId: req[options.auth.model].id
                },
                limit: 1
              }).then(function (data) {
                return data == 1
              }).catch(function (err) {
                res.database(err)
              });
            },

            /**
             * Checks for roles
             *
             * @param {Array<String>} roles Which roles
             * @returns {Promise.<Boolean>}
             */
            hasRoles: function (roles) {
              return Role.count({
                where: {
                  code: {
                    $in: roles
                  },
                  ownerId: req[options.auth.model].id
                }
              }).then(function (data) {
                return data == roles.length
              }).catch(function (err) {
                res.database(err)
              });
            }
          };

          if (req.originalUrl == '/login' || req.originalUrl == '/register' || req.originalUrl == '/') {
            logger.debug('Authentication: request passed by default allowed routes');
            return next();
          }

          var allowResult = allow.some(function (r) {
            return r.test(req.originalUrl);
          });

          if (allowResult) {
            logger.debug('Authentication: request passed by allow');
            return next();
          }

          var requiredRoles = [];
          routes.forEach(function (r) {
            if (r.match.test(req.originalUrl)) {
              requiredRoles.push(r.role);
            }
          });

          var userToken = req.get('token');
          if (userToken == undefined) return res.unauthorized('You entered an area that requires authorization. Please send token in headers');
          if (!/^[a-f0-9]{32}$/.test(userToken)) return res.unauthorized('Invalid token code');

          Token.find({
            include: [_gluon_auth_model],
            where: {code: userToken}
          }).then(function (token) {
            if (token == null || token.expire < new Date) return res.expiredToken('probably your token has been removed, please take new one');
            token.expire = Token.defaultExpire();
            token.save();

            req[options.auth.model] = token[_gluon_auth_model.name];
            if (requiredRoles.length == 0) {
              next();
            } else {
              Role.count({
                where: {
                  ownerId: req[options.auth.model].id,
                  code: {
                    $in: requiredRoles
                  }
                }
              }).then(function (count) {
                if (count != requiredRoles.length) return res.unauthorized('You do not have right roles to use this service');
                next();
              });
            }
          }).catch(function (err) {
            res.database(err)
          });
        });
      }
    } else {
      logger.log('Authentication: authentication and role management disabled!');
    }
  }
  else if (options.auth && options.auth.base == "token-redis") {
    options.auth = Object.assign(authDefaults, options.auth);
    if (options.auth.disable == false || options.auth.disable == undefined || options.auth.disable == "false") {
      var authLocation = path.resolve(process.cwd(), options.dir, options.models, options.auth.model);
      try {
        global._gluon_auth_model = require(authLocation);
        global._gluon_auth_expire = options.auth.expire;
      } catch (e) {
        logger.error('Auth model is not exist in {0}', authLocation);
      }

      if (global._gluon_auth_model) {
        var allow = options.auth.allow ? options.auth.allow.map(function (route, i) {
          return new RegExp(route.indexOf("regexp:") == 0 ? route.substring(7) : "^" + route.replace(/:[^\\/]+/g, '.*'));
        }) : [];

        var routes = Object.keys(options.auth.routes || {}).map(function (route, i) {
          return {
            match: new RegExp(route.indexOf("regexp:") == 0 ? route.substring(7) : "^" + route.replace(/:[^\\/]+/g, '.*')),
            role: options.auth.routes[route]
          };
        });

        var md5 = require('js-md5');
        var redisClient = require('./redis');

        app.use(function (req, res, next) {
          //noinspection JSValidateJSDoc
          req.auth = {
            /**
             * Creates a token for model
             * @param {Model} model
             * @returns {Promise.<TResult>}
             */
            login: function (model) {
              var tokenCode = 'token-' + md5(new Date().toString()) + md5(model.id + Math.random().toString());
              var redisExpire = (global._gluon_auth_expire) >>> 0; //redis using seconds not milliseconds
              return redisClient.setAsync(tokenCode, JSON.stringify(model.toJSON())).then(function () {
                return redisClient.expireAsync(tokenCode, redisExpire);
              }).then(function () {
                return {code: tokenCode, expire: new Date(Date.now() + redisExpire * 1000)};
              });
            },

            update: function (model) {
              var tokenCode = req.get('token');
              var redisExpire = (global._gluon_auth_expire) >>> 0; //redis using seconds not milliseconds
              return redisClient.setAsync(tokenCode, JSON.stringify(model.toJSON())).then(function () {
                return redisClient.expireAsync(tokenCode, redisExpire);
              }).then(function () {
                return {code: tokenCode, expire: new Date(Date.now() + redisExpire * 1000)};
              });
            },


            /**
             * Removes token from owner
             * @returns {Promise.<TResult>}
             */
            logout: function () {
              return redisClient.delAsync(req.get('token'));
            }
          };

          if (req.originalUrl == '/login' || req.originalUrl == '/register' || req.originalUrl == '/') {
            logger.debug('Authentication: request passed by default allowed routes');
            return next();
          }

          var allowResult = allow.some(function (r) {
            return r.test(req.originalUrl);
          });

          if (allowResult) {
            logger.debug('Authentication: request passed by allow');
            return next();
          }

          var userToken = req.get('token');
          if (userToken == undefined) return res.unauthorized('You entered an area that requires authorization. Please send token in headers');
          if (!/^token-[a-f0-9]{64}$/.test(userToken)) return res.unauthorized('Invalid token code');

          redisClient.getAsync(userToken).then(function (model) {
            if (model == null) return res.unauthorized('Invalid token code');
            var redisExpire = (global._gluon_auth_expire) >>> 0; //redis using seconds not milliseconds
            redisClient.expireAsync(userToken, redisExpire);

            req[options.auth.model] = JSON.parse(model);
            next();
          }).catch(function (err) {
            res.database(err)
          });
        });
      }
    } else {
      logger.log('Authentication: authentication and role management disabled!');
    }
  } else {
    logger.debug('Authentication: System ignored. Cause: {0}', options.auth ? 'authentication base invalid, only token supported' : 'no auth configuration');
  }


  logger.debug('Folder control unit loading..');
  requireDir(options.dir, options.models, function () {
    requireDir(options.dir, options.routes, function (files) {
      routeLoader(files, app);

      logger.debug('Gluon ready!');
      if (options.ready) options.ready(app, logger);

      if (options.listen != undefined) {
        if (typeof options.listen == 'object') {
          app.listen(process.env.port || options.listen.port, options.listen.ip);
          logger.log('Listener started on {2}:{0 def(1)}'.green, process.env.port, options.listen.port, options.listen.ip);
        } else if (typeof options.listen == 'number') {
          app.listen(process.env.port || options.listen);
          logger.log('Listener started on 0.0.0.0:{0 def(1)}'.green, process.env.port, options.listen);
        }
      }
    });
  });

  return app;
}

/**
 * Gluon router
 * @param {String} [location] Override location?
 * @param {Boolean} [ignore=false] Ignore me when routes binding
 * @param {Boolean} [mergeParams=true] Should i merge params
 * @returns {Router}
 */
Gluon.router = function (location, ignore, mergeParams) {
  if (mergeParams == undefined) mergeParams = true;

  var router = express.Router({mergeParams: mergeParams});
  router.location = location;
  router.ignore = ignore;
  return router;
};

module.exports = Gluon;